### ISSUE TITLE
100% CPU usage and memory leaks when converting inconsistent line endings in big XML files

### ISSUE THREAD

**Issue by VadimMuhtarov** [coteditor/CotEditor#1406]
### Description

The program hangs for an indefinite period of time with 100% CPU usage and high memory usage when parsing the big XML files. One time it used over 40 GB of virtual memory.

### To Reproduce

1. Open big XML file, I tried this (127 MB): http://aiweb.cs.washington.edu/research/projects/xmltk/xmldata/data/dblp/dblp.xml 
2. Press "Convert" inconsistent line endings

### Expected behavior

File is opened and successfully parsed 

### CotEditor version

4.4.2 (544) App Store

### macOS version

13.0.1 

### Additional context

MacBook Air M2

**Comment by VadimMuhtarov** [comment]
Log:
```
Date/Time:        2022-12-05 11:54:03.455 +0200
End time:         2022-12-05 11:55:33.496 +0200
OS Version:       macOS 13.0.1 (Build 22A400)
Architecture:     arm64e
Report Version:   40
Incident Identifier: E3A3D1BD-572D-4641-B030-F00472B4ABB8
Share With Devs:  Yes

Data Source:      Microstackshots
Shared Cache:     2D053604-1CB6-3821-A8DF-360EB2EB519B slid base address 0x198d9c000, slide 0x18d9c000

Command:          CotEditor
Path:             /Applications/CotEditor.app/Contents/MacOS/CotEditor
Identifier:       com.coteditor.CotEditor
Version:          4.4.2 (544)
Team ID:          HT3Z3A72WZ
App Item ID:      1024640650
App External ID:  853683554
Is First Party:   No
Resource Coalition ID: 1284
Architecture:     arm64
Parent:           UNKNOWN [1]
PID:              6987

Event:            cpu usage
Action taken:     none
CPU:              90 seconds cpu time over 90 seconds (100% cpu average), exceeding limit of 50% cpu over 180 seconds
CPU limit:        90s
Limit duration:   180s
CPU used:         90s
CPU duration:     90s
Duration:         90.04s
Duration Sampled: 85.32s (process was unresponsive for 37 seconds before sampling)
Steps:            46

Hardware model:   Mac14,2
Active cpus:      8
HW page size:     16384
VM page size:     16384

Advisory levels:  Battery -> 3, User -> 2, ThermalPressure -> 0, Combined -> 2
Free disk space:  103.32 GB/228.27 GB, low space threshold 3072 MB
Vnodes Available: 74.25% (89922/121100)

Preferred User Language: en-UA, uk-UA
Country Code:     UA
Keyboards:        U.S., Russian, Ukrainian-PC
OS Cryptex File Extents: 2050

Heaviest stack for the target process:
  42  start + 2544 (dyld + 24144) [0x198e3fe50]
  42  ??? (CotEditor + 20528) [0x1020e9030]
  42  NSApplicationMain + 880 (AppKit + 16472) [0x19c46b058]
  42  -[NSApplication run] + 500 (AppKit + 183316) [0x19c493c14]
  42  -[NSApplication _handleEvent:] + 60 (AppKit + 3895620) [0x19c81e144]
  42  -[NSApplication(NSEvent) sendEvent:] + 1920 (AppKit + 1462420) [0x19c5cc094]
  42  -[NSWindow(NSEventRouting) sendEvent:] + 284 (AppKit + 1466204) [0x19c5ccf5c]
  42  -[NSWindow(NSEventRouting) _reallySendEvent:isDelayedEvent:] + 1984 (AppKit + 1468656) [0x19c5cd8f0]
  42  -[NSWindow(NSEventRouting) _handleMouseDownEvent:isDelayedEvent:] + 3420 (AppKit + 1952992) [0x19c643ce0]
  42  -[NSControl mouseDown:] + 448 (AppKit + 1958380) [0x19c6451ec]
  42  -[NSButtonCell trackMouse:inRect:ofView:untilMouseUp:] + 488 (AppKit + 1961248) [0x19c645d20]
  42  -[NSCell trackMouse:inRect:ofView:untilMouseUp:] + 144 (AppKit + 1961576) [0x19c645e68]
  42  NSControlTrackMouse + 1480 (AppKit + 1963100) [0x19c64645c]
  42  -[NSButtonCell _sendActionFrom:] + 88 (AppKit + 1973912) [0x19c648e98]
  42  -[NSCell _sendActionFrom:] + 204 (AppKit + 1974132) [0x19c648f74]
  42  __26-[NSCell _sendActionFrom:]_block_invoke + 100 (AppKit + 1974348) [0x19c64904c]
  42  -[NSControl sendAction:to:] + 72 (AppKit + 1974536) [0x19c649108]
  42  -[NSApplication(NSResponder) sendAction:to:from:] + 440 (AppKit + 1975024) [0x19c6492f0]
  42  NSWindowEndWindowModalSession + 284 (AppKit + 8516616) [0x19cc86408]
  42  __54-[NSAlert beginSheetModalForWindow:completionHandler:]_block_invoke_2 + 40 (AppKit + 3844076) [0x19c8117ec]
  42  ??? (CotEditor + 267920) [0x102125690]
  42  ??? (CotEditor + 267508) [0x1021254f4]
  42  ??? (CotEditor + 253872) [0x102121fb0]
  42  ??? (CotEditor + 1039072) [0x1021e1ae0]
  42  -[NSMutableString(NSRegularExpressionSupport) _replaceOccurrencesOfRegularExpressionPattern:withTemplate:options:range:] + 220 (Foundation + 1132056) [0x19a207618]
  42  -[NSRegularExpression(NSReplacement) replaceMatchesInString:options:range:withTemplate:] + 240 (Foundation + 710584) [0x19a1a07b8]
  42  -[NSConcreteTextStorage replaceCharactersInRange:withString:] + 100 (UIFoundation + 238140) [0x19d3a923c]
  42  __NSConcreteTextStorageLockedForwarding + 88 (UIFoundation + 108624) [0x19d389850]
  42  -[NSTextStorage endEditing] + 88 (UIFoundation + 238408) [0x19d3a9348]
  42  -[NSTextStorage processEditing] + 220 (UIFoundation + 191320) [0x19d39db58]
  42  -[NSNotificationCenter postNotificationName:object:userInfo:] + 88 (Foundation + 38700) [0x19a0fc72c]
  42  _CFXNotificationPost + 708 (CoreFoundation + 293804) [0x19920fbac]
  42  _CFXRegistrationPost + 440 (CoreFoundation + 1130428) [0x1992dbfbc]
  42  ___CFXRegistrationPost_block_invoke + 88 (CoreFoundation + 1130612) [0x1992dc074]
  42  __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 128 (CoreFoundation + 484728) [0x19923e578]
  42  thunk for @escaping @callee_guaranteed @Sendable (@in_guaranteed Notification) -> () + 76 (Foundation + 3538228) [0x19a452d34]
  42  closure #1 in Notification.Subscription.init(_:_:_:_:) + 264 (Foundation + 3539812) [0x19a453364]
  42  Publishers.CompactMap.Inner.receive(_:) + 436 (Combine + 553780) [0x1a8e21334]
  42  protocol witness for Subscriber.receive(_:) in conformance Subscribers.Sink<A, B> + 24 (Combine + 122940) [0x1a8db803c]
  42  Subscribers.Sink.receive(_:) + 96 (Combine + 121024) [0x1a8db78c0]
  42  ??? (CotEditor + 1048164) [0x1021e3e64]
  42  ??? (CotEditor + 1045528) [0x1021e3418]
  14  _platform_memmove + 280 (libsystem_platform.dylib + 16360) [0x199197fe8]


Powerstats for:   CotEditor [6987]
UUID:             A6BC995A-DB77-3A4F-9908-E9ACD257E41C
Path:             /Applications/CotEditor.app/Contents/MacOS/CotEditor
Identifier:       com.coteditor.CotEditor
Version:          4.4.2 (544)
Team ID:          HT3Z3A72WZ
App Item ID:      1024640650
App External ID:  853683554
Is First Party:   No
Resource Coalition ID: 1284
Architecture:     arm64
Parent:           UNKNOWN [1]
UID:              501
Footprint:        882.56 MB -> 796.92 MB (-85.64 MB)
Start time:       2022-12-05 11:54:07.121 +0200
End time:         2022-12-05 11:55:32.438 +0200
Num samples:      46 (100%)
Note:             Unresponsive for 37 seconds before sampling
Primary state:    29 samples Non-Frontmost App, Non-Suppressed, User mode, Effective Thread QoS User Interactive, Requested Thread QoS User Interactive, Override Thread QoS Unspecified
User Activity:    0 samples Idle, 46 samples Active
Power Source:     0 samples on Battery, 46 samples on AC
  42  start + 2544 (dyld + 24144) [0x198e3fe50]
    42  ??? (CotEditor + 20528) [0x1020e9030]
      42  NSApplicationMain + 880 (AppKit + 16472) [0x19c46b058]
        42  -[NSApplication run] + 500 (AppKit + 183316) [0x19c493c14]
          42  -[NSApplication _handleEvent:] + 60 (AppKit + 3895620) [0x19c81e144]
            42  -[NSApplication(NSEvent) sendEvent:] + 1920 (AppKit + 1462420) [0x19c5cc094]
              42  -[NSWindow(NSEventRouting) sendEvent:] + 284 (AppKit + 1466204) [0x19c5ccf5c]
                42  -[NSWindow(NSEventRouting) _reallySendEvent:isDelayedEvent:] + 1984 (AppKit + 1468656) [0x19c5cd8f0]
                  42  -[NSWindow(NSEventRouting) _handleMouseDownEvent:isDelayedEvent:] + 3420 (AppKit + 1952992) [0x19c643ce0]
                    42  -[NSControl mouseDown:] + 448 (AppKit + 1958380) [0x19c6451ec]
                      42  -[NSButtonCell trackMouse:inRect:ofView:untilMouseUp:] + 488 (AppKit + 1961248) [0x19c645d20]
                        42  -[NSCell trackMouse:inRect:ofView:untilMouseUp:] + 144 (AppKit + 1961576) [0x19c645e68]
                          42  NSControlTrackMouse + 1480 (AppKit + 1963100) [0x19c64645c]
                            42  -[NSButtonCell _sendActionFrom:] + 88 (AppKit + 1973912) [0x19c648e98]
                              42  -[NSCell _sendActionFrom:] + 204 (AppKit + 1974132) [0x19c648f74]
                                42  __26-[NSCell _sendActionFrom:]_block_invoke + 100 (AppKit + 1974348) [0x19c64904c]
                                  42  -[NSControl sendAction:to:] + 72 (AppKit + 1974536) [0x19c649108]
                                    42  -[NSApplication(NSResponder) sendAction:to:from:] + 440 (AppKit + 1975024) [0x19c6492f0]
                                      42  NSWindowEndWindowModalSession + 284 (AppKit + 8516616) [0x19cc86408]
                                        42  __54-[NSAlert beginSheetModalForWindow:completionHandler:]_block_invoke_2 + 40 (AppKit + 3844076) [0x19c8117ec]
                                          42  ??? (CotEditor + 267920) [0x102125690]
                                            42  ??? (CotEditor + 267508) [0x1021254f4]
                                              42  ??? (CotEditor + 253872) [0x102121fb0]
                                                42  ??? (CotEditor + 1039072) [0x1021e1ae0]
                                                  42  -[NSMutableString(NSRegularExpressionSupport) _replaceOccurrencesOfRegularExpressionPattern:withTemplate:options:range:] + 220 (Foundation + 1132056) [0x19a207618]
                                                    42  -[NSRegularExpression(NSReplacement) replaceMatchesInString:options:range:withTemplate:] + 240 (Foundation + 710584) [0x19a1a07b8]
                                                      42  -[NSConcreteTextStorage replaceCharactersInRange:withString:] + 100 (UIFoundation + 238140) [0x19d3a923c]
                                                        42  __NSConcreteTextStorageLockedForwarding + 88 (UIFoundation + 108624) [0x19d389850]
                                                          42  -[NSTextStorage endEditing] + 88 (UIFoundation + 238408) [0x19d3a9348]
                                                            42  -[NSTextStorage processEditing] + 220 (UIFoundation + 191320) [0x19d39db58]
                                                              42  -[NSNotificationCenter postNotificationName:object:userInfo:] + 88 (Foundation + 38700) [0x19a0fc72c]
                                                                42  _CFXNotificationPost + 708 (CoreFoundation + 293804) [0x19920fbac]
                                                                  42  _CFXRegistrationPost + 440 (CoreFoundation + 1130428) [0x1992dbfbc]
                                                                    42  ___CFXRegistrationPost_block_invoke + 88 (CoreFoundation + 1130612) [0x1992dc074]
                                                                      42  __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 128 (CoreFoundation + 484728) [0x19923e578]
                                                                        42  thunk for @escaping @callee_guaranteed @Sendable (@in_guaranteed Notification) -> () + 76 (Foundation + 3538228) [0x19a452d34]
                                                                          42  closure #1 in Notification.Subscription.init(_:_:_:_:) + 264 (Foundation + 3539812) [0x19a453364]
                                                                            42  Publishers.CompactMap.Inner.receive(_:) + 436 (Combine + 553780) [0x1a8e21334]
                                                                              42  protocol witness for Subscriber.receive(_:) in conformance Subscribers.Sink<A, B> + 24 (Combine + 122940) [0x1a8db803c]
                                                                                42  Subscribers.Sink.receive(_:) + 96 (Combine + 121024) [0x1a8db78c0]
                                                                                  42  ??? (CotEditor + 1048164) [0x1021e3e64]
                                                                                    42  ??? (CotEditor + 1045528) [0x1021e3418]
                                                                                      14  _platform_memmove + 280 (libsystem_platform.dylib + 16360) [0x199197fe8]
                                                                                        6   <Frontmost App>
                                                                                      13  _platform_memmove + 536 (libsystem_platform.dylib + 16616) [0x1991980e8]
                                                                                        5   <Frontmost App>
                                                                                      8   ??? (CotEditor + 1047000) [0x1021e39d8]
                                                                                        2   <Frontmost App>
                                                                                      5   ??? (CotEditor + 1047016) [0x1021e39e8]
                                                                                      1   _platform_memmove + 544 (libsystem_platform.dylib + 16624) [0x1991980f0]
                                                                                      1   ??? (CotEditor + 1047012) [0x1021e39e4]
  4   start_wqthread + 0 (libsystem_pthread.dylib + 7704) [0x199164e18]
    3   <Kernel mode>
    1   <Frontmost App, Kernel mode>

  Binary Images:
           0x1020e4000 -        0x102413fff  com.coteditor.CotEditor 4.4.2 (544)   <A6BC995A-DB77-3A4F-9908-E9ACD257E41C>  /Applications/CotEditor.app/Contents/MacOS/CotEditor
           0x198e3a000 -        0x198ec44bf  dyld (1042.1)                         <DE46DD52-4994-3FD8-B4B4-E352A1A19354>  /usr/lib/dyld
           0x199163000 -        0x19916fffb  libsystem_pthread.dylib (514)         <886CACA0-5762-3640-8DB2-3FA3B911C062>  /usr/lib/system/libsystem_pthread.dylib
           0x199194000 -        0x19919bffb  libsystem_platform.dylib (288)        <78189670-2F47-332C-9A59-A210157AA97B>  /usr/lib/system/libsystem_platform.dylib
           0x1991c8000 -        0x19969ffff  com.apple.CoreFoundation 6.9 (1953.1) <F4FF83FC-E62C-30B4-B3A9-876C8A1FD595>  /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation
           0x19a0f3000 -        0x19ab2bfff  com.apple.Foundation 6.9 (1953.1)     <52B2F83F-586D-32A2-8B66-CE651400AC8C>  /System/Library/Frameworks/Foundation.framework/Versions/C/Foundation
           0x19c467000 -        0x19d36efff  com.apple.AppKit 6.9 (2299)           <AF9F6891-70AD-3C26-AF08-B747344892D2>  /System/Library/Frameworks/AppKit.framework/Versions/C/AppKit
           0x19d36f000 -        0x19d513fff  com.apple.UIFoundation 1.0 (855)      <6B5DCD6A-9C35-30E4-8A02-6A73A39D985E>  /System/Library/PrivateFrameworks/UIFoundation.framework/Versions/A/UIFoundation
           0x1a8d9a000 -        0x1a8ee9ff7  com.apple.combine 1.0 (310)           <DE89D083-235E-3A21-97EE-69E248E6BF15>  /System/Library/Frameworks/Combine.framework/Versions/A/Combine

```

**Comment by 1024jp** [comment]
Thank you for the report.
I confirmed the issue.
However, I think it'll take time to solve it. Let me have time.

**Comment by 1024jp** [comment]
I suppose I succeed in significantly reduce the calculation cost on the normalization of inconsistent line endings.
The change will be shipped with CotEditor 4.7.0, which will be released in this December.

Again, thank you for the report.
